swxsoc/util/util.py

```diff
@@ -202,3 +211,225 @@ def parse_science_filename(filepath: str) -> dict:
+    start_time : str
+        The start time in ISO format.
+    end_time : str
+        The end time in ISO format.
+
+    Returns
+    -------
+    list
+        A list of prefixes.
+    """
+    current_time = start_time
+    prefixes = []
+
+    while current_time <= end_time:
+        prefix = f"{level}/{current_time.year}/{current_time.month:02d}/"
+        prefixes.append(prefix)
+        current_time += relativedelta(months=1)
+
+    return prefixes
+
+
+def get_latest_dependent_file(
+    instrument: str,
+    level: str,
+    start_time: str,
+    end_time: str,
+    mode: str = None,
+    test: bool = None,
+    version: str = None,
+    descriptor: str = None,
+    use_s3: bool = True,
+) -> Path:
+    """
+    Retrieves the latest dependent file based on the provided criteria.
+
+    Parameters
+    ----------
+    instrument : str
+        The instrument name.
+    level : str
+        The data level.
+    start_time : str
+        The start time in ISO format.
+    end_time : str
+        The end time in ISO format.
+    mode : str, optional
+        The mode to filter by.
+    test : bool, optional
+        The test flag to filter by.
+    version : str, optional
+        The version to filter by.
+    descriptor : str, optional
+        The descriptor to filter by.
+    use_s3 : bool, optional
+        Whether to get file from S3 or fallback to SPDF (default is True).
+
+    Returns
+    -------
+    Path
+        The path to the latest dependent file, or None if no matching file is found.
+    """
+    start_time = parse_date(start_time)
+    end_time = parse_date(end_time)
+
+    bucket_prefix = "" if os.getenv("LAMBDA_ENVIRONMENT") == "PRODUCTION" else "dev-"
+    mission_name = swxsoc.config["mission"]["mission_name"]
+    bucket_name = (
+        f"{bucket_prefix}{mission_name}-{instrument}"
+        if instrument in swxsoc.config["mission"]["inst_names"]
+        else f"{bucket_prefix}{mission_name}-ancillary"
+    )
+
+    prefixes = generate_prefixes(level, start_time, end_time)
+    all_files = []
+
+    try:
+        for prefix in prefixes:
+            all_files.extend(list_files_in_s3(bucket_name, prefix))
+    except Exception:
+        use_s3 = False
+
+    all_files.sort(key=lambda x: parse_science_filename(x)["time"], reverse=True)
+
+    for file in all_files:
+        try:
+            science_file_dict = parse_science_filename(file)
+            file_time = science_file_dict["time"]
+
+            if start_time <= file_time <= end_time:
+                if filter_science_file(
+                    science_file_dict,
+                    instrument,
+                    mode,
+                    test,
+                    version,
+                    descriptor,
+                ):
+                    if not use_s3:
+                        return download_file_from_spdf(file, "/tmp")
+                    else:
+                        return download_file_from_s3(bucket_name, file, "/tmp")
```

Comment:
Do you think it would be better to return a file id or path here compared to downloading the whole file? That way the filtering and querying is separated from the downloading.